docs/tracing.md

```diff
@@ -0,0 +1,281 @@
+
+### Tracing Levels
+
+Tracing levels refer to the severity of the span or event being created. If a subscribe records tracing data at level `INFO` then it records all tracing data at or below the `INFO` level. That is a lower number indicates a higher priority.
+
+These can be one of:
+
+|Priority|Level|Usage|Example|
+|---|---|---|---|
+|1|Error|When a state arises which either results in the termination of the program or significantly affects its smooth running.|An incorrect configuration is given during start-up.|
+|2|Warn|When a state arises which prevents the program doing its job properly, but could be recoverable in the future.|A corrupted or unidentifiable Kafka message, or series of invalid Run commands.|
+|3|Info|To indicate the normal running of the program at a coarse-grained level, as well as collate spans from other components.|The `Run` span in the `nexus-writer` tool which links to all traces relevant to the same run.|
+|4|Debug|For events and spans which may assist in debugging issues. Generally low priority.|Some details of event formation|
+|5|Trace|For the most fine-grained spans and traces. Generally each and every function should be instrumented at the trace level.|Minutae of event formation, IO, or disk write functions|
+
+## Standards
+
+### Span and Events
+
+In general we prefer spans over events as spans allow for an interval of execution to be recorded.
+Spans should be structured with appropriate fields rather than relevant information being recorded by events.
+Fields allow information to be recorded as key/value pairs that can be accessed later.
+
+### Instrument Macro for Functions
+
+Use of the `#[instrument]` macro is preferred over defining spans directly. If necessary, the `name` field can be overridden by the syntax `#[instrument(name = ...)]`.
+
+The syntax `#[instrument(skip_all)]` should be used to ignore function parameters by default. These can be added in as `fields` as necessary.
+
+Include the line `use tracing::instrument;` to bring the macro into scope.
+By default `#[instrument]` creates an `INFO` level span, this can be overridden by the syntax `#[instrument(level = ...)]`.
+
+Spans which should only be collected by OpenTelemetry should be given target `otel` by the syntax `#[instrument(target = "otel")]`. Use this to avoid spans or events being sent to stdout.
+
+Every function that can fail should be instrumented (i.e. that has return type `Result<>`), and should have use `#[instrument(err(level = "WARN"))]` or `#[instrument(err(level = ERROR))]` depending on the type of error.
+
+### Tracing and Parallel Execution
```

Comment:
For the sake of clarity, fully qualify things that are not in the standard prelude in the code examples (including traits imports).

Reply:
Does this mean things like using
`rayon::IntoParallelRefIterator::par_iter(&vector)` rather than `vector.par_iter()`?

Reply:
No, but include trait imports.